src/main/java/dev/blaauwendraad/masker/json/KeyMatcher.java

```diff
@@ -313,211 +351,311 @@ && isEncodedCharacter(bytes, i, endIndex)) {
-            PreInitTrieNode child = children.get(b);
+        @Nullable
+        PreInitTrieNode child(byte b) {
+            PreInitTrieNode child = childrenLowercase.get(b);
             if (child != null) {
                 return child;
             }
-            return childrenUpper.get(b);
+            return childrenUppercase.get(b);
         }
 
         /**
          * Adds a new child to the trie. When case-insensitivity is enabled this must represent the
          * lower-case byte, otherwise is just a byte in original case.
          */
-        void add(Byte b, PreInitTrieNode child) {
-            children.put(b, child);
+        void addLowercase(Byte b, PreInitTrieNode child) {
+            childrenLowercase.put(b, child);
         }
 
         /**
          * Adds a child using an equivalent upper-case byte of the child already inserted. When
          * case-insensitivity is enabled this must represent the upper-case byte, otherwise must not
          * be called.
          */
-        void addUpper(Byte b, PreInitTrieNode child) {
-            childrenUpper.put(b, child);
+        void addUppercase(Byte b, PreInitTrieNode child) {
+            childrenUppercase.put(b, child);
+        }
+    }
+
+    /**
+     * A helper class for matching against the (radix) trie node.
+     * <p>During matching the node keeps transient matching state - reference to the node {@link #tempNode} and
+     * the {@link #tempPrefixIndex} for keeping track of how much of the common prefix has been matched already.
+     * During matching, this helper class changes the
+     *
+     * <p> After the matching, the node can be reset to the initial state using {@link #reset()}.
+     */
+    static class StatefulRadixTrieNode {
+        private final RadixTrieNode node;
+        private final int prefixIndex;
+
+        private RadixTrieNode tempNode;
+        private int tempPrefixIndex;
+
+        StatefulRadixTrieNode(RadixTrieNode node, int prefixIndex) {
+            this.node = this.tempNode = node;
+            this.prefixIndex = this.tempPrefixIndex = prefixIndex;
+        }
+
+        StatefulRadixTrieNode(StatefulRadixTrieNode node) {
+            this(node.tempNode, node.tempPrefixIndex);
+        }
+
+        /**
+         * Returns a child of this node representing the provided byte value or {@code null} if such child does not exist.
+         *
+         * <p>This mutates the current state
+         * prefix matches the byte value, return the existing radix trie node with increased prefix index. Otherwise,
+         * looks if a child node matches the byte value. If neither match, returns {@code null}.
+         *
+         * @param byteValue the "next" byte value to find the matching radix trie node for
+         * @return the current radix trie node or a child of it, if either of them match. Otherwise, returns
+         * {@code null}
+         */
+        boolean performChildLookup(byte byteValue) {
```

Comment:
The name is a bit confusing as I would expect the method to return a looked up child. How about calling it "descent"? This would indicate that the method changes the state of the object.